src/libraries/System.Linq.Expressions/src/System/Linq/Expressions/LambdaExpression.cs

```diff
@@ -102,6 +102,21 @@ internal virtual int ParameterCount
             }
         }
 
+        [DynamicDependency("Compile()", typeof(LambdaExpression))]
+        [DynamicDependency("Compile()", typeof(Expression<>))]
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
+            Justification = "The 'Compile' methods will be preserved by the DynamicDependency.")]
+        internal static MethodInfo GetCompileMethod(Type lambdaExpressionType)
+        {
+            Debug.Assert(lambdaExpressionType.IsAssignableTo(typeof(LambdaExpression)));
+
+            MethodInfo result = lambdaExpressionType.GetMethod("Compile", Type.EmptyTypes)!;
+            Debug.Assert(result.DeclaringType == typeof(LambdaExpression) ||
```

Comment:
Would it make sense to just rewrite this in explicit terms into something along the lines of
```
if (lambdaExpressionType == typeof(LambdaExpression)
    return typeof(LambdaExpression).GetMethod(...)
else
    return lambdaExpressionType.GetMemberWithSameMetadataDefinition(typeof(Expression<>).GetMethod(...));
```

Of course we don't have GetMemberWithSameMetadataDefinition right now, but that would be the end goal (I consider the other places where we put DynamicDependency because we don't have this API temporary crutches).

Asserts are okay way to guard trimming safety, but compile time enforcement trumps asserts and this structure has guaranteed same behavior before/after trimming.

Reply:
Yes, I've added this to the places to change once we get that API. #45771.

I refactored this method a bit which should help here.